widget/slider.go

```diff
@@ -0,0 +1,289 @@
+	padLen := theme.Padding()
+	sideLen := padLen * standardScale
+	activeOffset, thumbOffset := s.moveSlide(sideLen)
+
+	var trackPos, activePos, thumbPos fyne.Position
+	var trackSize, activeSize fyne.Size
+
+	if s.slider.Vertical {
+		trackPos = fyne.NewPos(size.Width/2, 0)
+		activePos = fyne.NewPos(trackPos.X, activeOffset)
+
+		trackSize = fyne.NewSize(padLen, size.Height)
+		activeSize = fyne.NewSize(padLen, trackSize.Height-activeOffset)
+
+		thumbPos = fyne.NewPos(
+			trackPos.X-(sideLen-trackSize.Width)/2, thumbOffset)
+	} else {
+		trackPos = fyne.NewPos(0, size.Height/2)
+		activePos = trackPos
+
+		trackSize = fyne.NewSize(size.Width, padLen)
+		activeSize = fyne.NewSize(activeOffset, padLen)
+
+		thumbPos = fyne.NewPos(
+			thumbOffset, trackPos.Y-(sideLen-trackSize.Height)/2)
+	}
+
+	s.track.Move(trackPos)
+	s.track.Resize(trackSize)
+
+	s.active.Move(activePos)
+	s.active.Resize(activeSize)
+
+	s.thumb.Move(thumbPos)
+	s.thumb.Resize(fyne.NewSize(sideLen, sideLen))
+}
+
+// MinSize calculates the minimum size of a slider widget
+func (s *sliderRenderer) MinSize() fyne.Size {
+	s1, s2 := minLongSide, theme.Padding()*standardScale
+	if s.slider.Vertical {
+		return fyne.NewSize(s2, s1)
+	}
+	return fyne.NewSize(s1, s2)
+}
+
+func (s *sliderRenderer) BackgroundColor() color.Color {
+	return theme.BackgroundColor()
+}
+
+func (s *sliderRenderer) Destroy() {
+}
+
+func (s *sliderRenderer) Objects() []fyne.CanvasObject {
+	return s.objects
+}
+
+func (s *sliderRenderer) moveSlide(diameter int) (int, int) {
+	w := s.slider
+	t := s.track.Size()
+	ratio := (w.Value - w.Min) / (w.Max - w.Min)
+	if w.Vertical {
+		y := float64(t.Height) - (ratio * float64(t.Height))
+		return int(y), int(y) - int((1-ratio)*float64(diameter))
+	}
+	x := ratio * float64(t.Width)
+	return int(x), int(x) - int(ratio*float64(diameter))
+}
+
+func checkStep(s, max, min float64) {
+	// make sure there is a positive step and it is less than
+	// the maximum value
+	if s <= 0 {
+		fyne.LogError("Step is less than or equal to zero.", nil)
+	}
+	if s*s >= (max-min)*(max-min) {
+		fyne.LogError("Step is greater than or equal to range.", nil)
+	}
+}
+
+func checkMinMax(val, min, max float64) (float64, float64) {
+	// sort the values to ensure correct order
+	if val < min {
+		fyne.LogError("Value is less minimum value.", nil)
+		min = val
+	}
+	if val > max {
+		fyne.LogError("Value is greater than maximum value.", nil)
+		max = val
+	}
+	if min == max {
+		fyne.LogError("Minimum value equals maximum value.", nil)
+		min--
+		max++
+	}
+	if min > max {
+		fyne.LogError("Minimum value is greater than maximum value.", nil)
+		return max, min
+	}
+	return min, max
+}
+
+// NewSlider returns a basic slider.
+// value - the initial slider value
+// min   - the minimum value in the range
+// max   - the maximum value in the range
+// step  - the incremental step count (needs to be positive and < max)
+func NewSlider(value, min, max, step float64) *Slider {
```

Comment:
With all of these parameters the constructor is a little confusing.
Remember that people can choose to use the struct version .i.e:
```go
&Slider{
    Value: 1,
    Min: 0,
    Max: 5,
    Step: 0.5,
}
```

So we should consider what is truly the minimal values for a constructor?
Perhaps min and max is enough?
I would imagine that it would be most common for step to be 1 and that a sensible default value would be the same as min?
This is open for discussion, but passing 4 parameters like this will lead to confusion as docs must be consulted to know what to do...

Reply:
I am completely for making the constructor more minimal, personally I like minimal interfaces. In fact, I am cringing right now thinking about the gl library.  I am on board with your ideas about the interface and the default values, and to be honest I am tempted to mimic the QT interface and only pass the orientation.

```go
type Orientation int

const (
	Horizontal Orientation = 0
	Vertical   Orientation = 1
)

type Slider struct {
	baseWidget

	Value float64
	Min   float64
	Max   float64
	Step  float64

	Orientation Orientation
	OnChanged   func(float64)
}

func NewSlider(orientation Orientation) *Slider {
	return &Slider{
		Value:       0,
		Min:         0,
		Max:         10,
		Step:        1,
		Orientation: orientation,
	}
}

```
Eliminate the error checking functions and make these changes?

Reply:
On the comparison with Qt I don't think I agree. Most scrollbars will be horizontal (I would think?) and so requiring every one to specify seems overkill. In my mind it's far more likely that min and max would vary frequently? i.e. colours are 0-255 booleans are 0-1, scores 1-5, for example. I don't see the user interface often flowing in the vertical arrangement?

Reply:
Yeah, I can see the argument there, going back to the conversation [#389 ](https://github.com/fyne-io/fyne/issues/389)about values and ranges that are limited to resolution. A radial slider (knob) 😆 would allow you to have and infinite resolution inside a finite width and height. Weirdly enough the associated motion interacting with that type of widget is more suited to mobile I think,